### PR TITLE
fix(jans-fido2): widen metrics columns and restore missing config keys

### LINKED ISSUE #14687: fido2: metrics collect no data — undersized metrics columns, and 14 config keys missing from fido2-metrics.properties
**Describe the bug**

FIDO2 telemetry collects nothing. There are two independent defects behind this, each introduced by a different PR. Both must be fixed for the feature to work.

---

**Defect 1 — metrics columns are `VARCHAR(64)`, too small for a `User-Agent`**

Every attempt to write a raw metrics entry is rejected by the database:

```
08:17:18.983 ERROR [io.jans.fido2.service.metric.Fido2MetricsService] (Fido2MetricsService.java:69)
  Failed to store FIDO2 metrics entry: Failed to persist entry:
  'jansId=a5bd39d1-...,ou=fido2-metrics,o=jans'
Caused by: org.postgresql.util.PSQLException:
  ERROR: value too long for type character varying(64)
```

`jansFido2MetricsEntry` therefore stays empty, and the aggregation job — which runs correctly and on schedule — has nothing to summarise and writes no aggregation rows. To an operator this looks like a broken scheduler, but the scheduler is not at fault.

The write is fire-and-forget (`CompletableFuture.runAsync`, exception swallowed into `log.error`), so the passkey registration itself still succeeds and nothing surfaces to the user. The metric is simply lost.

**Defect 2 — two services cannot initialize at all**

`Fido2UserMetricsService` and `Fido2AnalyticsService` read their configuration from `fido2-metrics.properties` into `static final` fields. Fourteen of the keys they read no longer exist in that file, so static initialization throws and neither class can be loaded:

```
java.lang.ExceptionInInitializerError
Caused by: java.util.MissingResourceException: Can't find resource for bundle
  java.util.PropertyResourceBundle, key fido2.user.metrics.base.dn
```

Any subsequent use fails with `NoClassDefFoundError: Could not initialize class io.jans.fido2.service.metric.Fido2UserMetricsService`. This is invisible in the log: `MetricService.updateUserMetrics` (`MetricService.java:898`) catches `Exception`, but `ExceptionInInitializerError` extends `Error` and is not caught there.

---

**To Reproduce**

1. Install Janssen/Flex with FIDO2 on a PostgreSQL or MySQL backend
2. Confirm `fido2MetricsEnabled` and `fido2MetricsAggregationEnabled` are `true`
3. Register a passkey from a normal desktop browser, so a real `User-Agent` header (100–350 characters) is sent
4. Check the FIDO2 log — every registration/authentication produces `Failed to store FIDO2 metrics entry ... value too long for type character varying(64)`
5. Query the tables — both stay empty:
   ```sql
   SELECT count(*) FROM "jansFido2MetricsEntry";   -- 0, defect 1
   SELECT count(*) FROM "jansFido2UserMetrics";    -- 0, defect 2
   ```
6. Wait for the next hourly run. The log shows `Hourly aggregation completed for: <hour>`, but `jansFido2MetricsAggregation` remains empty

Defect 2 also reproduces without a server: write any test that instantiates `Fido2UserMetricsService` (for example with `@InjectMocks` under `MockitoExtension`) and it fails to initialize.

**Expected behavior**

Metrics entries are stored with `userAgent` and `deviceInfo` intact, per-user metrics are written to `jansFido2UserMetrics`, and the hourly aggregation produces a row for that window.

**Screenshots**

N/A — server-side, reproduced from logs and database queries.

**Desktop / Smartphone**

Not client-specific. Any browser reproduces defect 1, because every mainstream browser sends a `User-Agent` longer than 64 characters.

---

**Additional context**

**Root cause — defect 1**

Every `jansFido2Metrics*` attribute is declared in `jans_setup/schema/jans_schema.json` with syntax `1.3.6.1.4.1.1466.115.121.1.15`, which `static/rdbm/ldap_sql_data_type_mapping.json` maps to `VARCHAR(64)`. `static/rdbm/sql_data_types.json` is the supported per-attribute override file, but PR #12143 — which introduced the metrics model — added the attributes to the schema without adding any size override. Every column was created 64 characters wide, including ones that hold serialized JSON.

| Column | Typical size written |
|---|---|
| `jansFido2MetricsUserAgent` — raw `User-Agent` header | 100–350 chars |
| `jansFido2MetricsDeviceInfo` — `@JsonObject` blob that re-embeds the user agent | 100–700 chars |
| `jansFido2MetricsErrorReason` — raw `exception.getMessage()` | unbounded |

`jansFido2UserMetrics` has the same defect via `jansLastUserAgent`.

**Root cause — defect 2**

Commit `c898187a8f5` (PR #12934) rewrote `fido2-metrics.properties`, replacing 99 lines with 24. The rewrite dropped every `fido2.user.*` and `fido2.performance.*` key, but the classes consuming them were not updated.

| Class | Lines | Keys required | Present |
|---|---|---|---|
| `Fido2UserMetricsService` | 49–60 | 12 (`fido2.user.*`) | 0 |
| `Fido2AnalyticsService` | 50–53 | 4 (`fido2.user.*`, `fido2.performance.*`) | 0 |

`Fido2MetricsAggregationScheduler` reads the same bundle but is unaffected — it wraps the lookup in `try`/`catch` and falls back to hardcoded defaults (`Fido2MetricsAggregationScheduler.java:75-86`), which is why the aggregation scheduler keeps running normally.

**Impact**

FIDO2 telemetry has never collected data on any deployment built since these changes landed. Existing VM deployments will not self-heal on upgrade, since `jans-linux-setup` skips tables that already exist and emits no `ALTER COLUMN`; container and Kubernetes deployments correct the column widths automatically through the persistence loader's schema-diff step.

**Environment**

- Janssen/Flex with FIDO2, current build
- Persistence: PostgreSQL. MySQL is affected identically, reporting `Data too long for column`

---

**Proposed fix**

A fix is prepared and will follow in a PR:

1. **Column widths** — add the missing overrides to `static/rdbm/sql_data_types.json`: `VARCHAR(512)` for the user agent columns, `VARCHAR(1024)` for `errorReason`, `VARCHAR(256)`/`VARCHAR(128)` for the user and session identifiers, and `TEXT` for the serialized JSON blobs. Deliberately **not** `rdbm_json_column`, since the SQL ORM treats native JSON columns as multivalued and would silently change the stored shape from `{...}` to `[{...}]`, and the container migration for that conversion drops and recreates the column.

2. **Length guard in code** — shorten free-form values to the column width before persisting, so an unbounded input (raw exception text, client-controlled cookie or username) can never again cost the entire row. One `WARN` per entry naming the field and its original length; the value itself is never logged, since these fields are personal data. Applied to both write paths.

3. **Restore the missing keys** with the values they had in `39d742bbe4c`. Longer term these two classes should either tolerate a missing key the way the scheduler does, or move to the FIDO2 dynamic configuration — a missing entry in a packaged resource should not be able to take a service down at class-load time.

Also included: a documented one-time `ALTER TABLE` migration for existing VM installs, and unit tests covering the truncation boundaries, the surrogate-pair edge case, and a regression guard asserting that a typical 250-character user agent is stored **intact** rather than shortened.

### PR DESCRIPTION
**Pull request by imran-ishaq**
### Description

FIDO2 telemetry collected no data at all: every metrics write was rejected by
the database because the `jansFido2Metrics*` columns were created as
`VARCHAR(64)`, which cannot hold a browser `User-Agent`. Separately, 14 keys
that `Fido2UserMetricsService` and `Fido2AnalyticsService` read into static
final fields had been dropped from `fido2-metrics.properties`, so neither
class could initialize. This widens the affected columns, shortens free-form
values to the column width before persisting, and restores the missing keys.

### Target Issue
FIDO2 telemetry collects no data. Metrics writes are rejected with
"value too long for type character varying(64)" because the
`jansFido2Metrics*` columns were created at the schema default width,
and separately `Fido2UserMetricsService` and `Fido2AnalyticsService`
cannot initialize because 14 keys they read were dropped from
`fido2-metrics.properties`.

  
  
  
closes #14687

#### Implementation Details
  

-------------------
### Test and Document the changes
- [ ] Static code analysis has been run locally and issues have been fixed
- [ ] Relevant unit and integration tests have been added/updated
- [ ] Relevant documentation has been updated if any (i.e. user guides, installation and configuration guides, technical design docs etc)

Please check the below before submitting your PR. The PR will not be merged if there are no commits that start with `docs:` to indicate documentation changes or if the below checklist is not selected.
- [ ] **I confirm that there is no impact on the docs due to the code changes in this PR.**

## Summary by CodeRabbit

* **Bug Fixes**
  * Prevented oversized FIDO2 metric and user-metric values from exceeding storage limits.
  * Preserved valid, blank, and multibyte data safely, including expanded user-agent limits.
  * Added warnings for truncated fields without exposing sensitive values in logs.

* **Configuration**
  * Added metric storage, threshold, risk, and slow-operation settings.
  * Updated database definitions for FIDO2 and user metric fields.

* **Documentation**
  * Documented field limits, privacy-safe truncation, troubleshooting, and migration guidance.

* **Tests**
  * Added coverage for truncation, validation, device information, logging safety, and persistence flows.

### CODE REVIEW

**Comment by mo-auto** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/janssenproject/pr-checks/9bc2e348-5cae-4136-b929-558fda8d2ce7?refs=5b7354c6892f9e9d0e32f3f4dc88abaafc7e8fb9&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: docs/janssen-server/fido/passkey-telemetry.md [hunk 1]
```diff
@@ -214,6 +222,141 @@ is given below.
...
+
+**2. Aggregation.** The hourly job runs a few minutes past each hour and summarises the
+*previous completed* hour in UTC, keyed as `HOURLY_<yyyy-MM-dd-HH>`. If you registered at
+14:30 UTC, wait for the run just after 15:00 and look for the `14` bucket:
+
+```sql
+SELECT "jansId", "jansStartTime", "jansEndTime"
+FROM "jansFido2MetricsAggregation"
+WHERE "jansId" = 'HOURLY_2026-08-03-14';
+```
```
[the remark this answers is not in the document]
Comment (imran-ishaq) [reply docs/janssen-server/fido/passkey-telemetry.md]: Done

File: docs/janssen-server/fido/passkey-telemetry.md [hunk 2]
```diff
@@ -214,6 +222,141 @@ is given below.
...
+Note the **username** you registered with and the **UTC time** of the attempt. Every check
+below filters on them, so a row left over from earlier traffic cannot make a failed migration
+look successful.
+
+**1. Raw entries.** Both write paths should produce a row for that user:
+
+```sql
+-- registration/authentication events for the test account
+SELECT "jansFido2MetricsTimestamp", "jansFido2MetricsOperationType", "jansFido2MetricsStatus",
+       length("jansFido2MetricsUserAgent") AS ua_length
```
[the remark this answers is not in the document]
Comment (imran-ishaq) [reply docs/janssen-server/fido/passkey-telemetry.md]: Done

File: jans-fido2/server/src/main/java/io/jans/fido2/service/metric/MetricsFieldTruncation.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+        int cut = maxLength;
+        if (Character.isHighSurrogate(value.charAt(cut - 1))) {
+            cut--;
+        }
+
+        truncatedFields.add(fieldName + " (" + value.length() + " > " + maxLength + ")");
+
+        return value.substring(0, cut);
+    }
```
[the remark this answers is not in the document]
Comment (imran-ishaq) [reply jans-fido2/server/src/main/java/io/jans/fido2/service/metric/MetricsFieldTruncation.java:61]: Done with changes

File: jans-fido2/server/src/test/java/io/jans/fido2/service/metric/Fido2MetricsServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+
+        // When
+        fido2MetricsService.storeMetricsData(metricsData);
+
+        // Then
+        Fido2MetricsEntry entry = capturePersistedEntry();
+        assertNotNull(entry.getDeviceInfo());
+        assertEquals(Fido2MetricsConstants.MAX_LENGTH_USER_AGENT, entry.getDeviceInfo().getUserAgent().length());
+        assertEquals("Chrome", entry.getDeviceInfo().getBrowser(), "short members must be untouched");
+    }
```
[the remark this answers is not in the document]
Comment (imran-ishaq) [reply jans-fido2/server/src/test/java/io/jans/fido2/service/metric/Fido2MetricsServiceTest.java:197]: Done

**Review by yurem — APPROVED** [review]
